gridstatus/isone_api/isone_api.py

```diff
@@ -0,0 +1,379 @@
+from gridstatus.base import NoDataFoundException
+from gridstatus.decorators import support_date_range
+from gridstatus.gs_logging import logger as log
+
+# API base URL
+BASE_URL = "https://webservices.iso-ne.com/api/v1.1"
+
+# Default page size for API requests
+DEFAULT_PAGE_SIZE = 1000
+
+
+# See more info here: https://www.iso-ne.com/participate/support/web-services-data#loadzone
+ZONE_LOCATIONID_MAP = {
+    "NEPOOL AREA": 32,
+    ".H.INTERNALHUB": 4000,
+    ".Z.MAINE": 4001,
+    ".Z.NEWHAMPSHIRE": 4002,
+    ".Z.VERMONT": 4003,
+    ".Z.CONNECTICUT": 4004,
+    ".Z.RHODEISLAND": 4005,
+    ".Z.SEMASS": 4006,
+    ".Z.WCMASS": 4007,
+    ".Z.NEMASSBOST": 4008,
+}
+
+EXCLUDE_FROM_REALTIME_HOURLY_DEMAND = [
+    "NEPOOL AREA",
+    ".H.INTERNALHUB",
+]
+
+
+class ISONEAPI:
+    """
+    Class to authenticate with and make requests to the ISO New England API.
+
+    To authenticate, you need a username and password.
+
+    To register, create an account here: https://www.iso-ne.com/participate/applications-status-changes/access-software-systems#ws-api
+    """
+
+    default_timezone = "US/Eastern"
+
+    def __init__(
+        self,
+        sleep_seconds: float = 0.2,
+        max_retries: int = 3,
+    ):
+        self.username = os.getenv("ISONE_API_USERNAME")
+        self.password = os.getenv("ISONE_API_PASSWORD")
+
+        if not all([self.username, self.password]):
+            raise ValueError(
+                "Username and password must be provided or set as environment variables",
+            )
+
+        self.sleep_seconds = sleep_seconds
+        self.initial_delay = min(max(0.1, sleep_seconds), 60.0)
+        self.max_retries = min(max(0, max_retries), 10)
+
+    def _handle_end_date(
+        self,
+        date: pd.Timestamp,
+        end: pd.Timestamp | None,
+        days_to_add_if_no_end: int,
+    ) -> pd.Timestamp:
+        """
+        Handles a provided end date by either
+
+        1. Using the provided end date converted to the default timezone
+        2. Adding the number of days to the date and converting to the default timezone
+        """
+        if end:
+            end = utils._handle_date(end, tz=self.default_timezone)
+        else:
+            # Have to convert to UTC to do addition, then convert back to local time
+            # to avoid DST issues
+            end = (
+                (date.tz_convert("UTC") + pd.DateOffset(days=days_to_add_if_no_end))
+                .normalize()
+                .tz_localize(None)
+                .tz_localize(self.default_timezone)
+            )
+
+        return end
+
+    def make_api_call(
+        self,
+        url: str,
+        api_params: dict = None,
+        parse_json: bool = True,
+        verbose: bool = False,
+    ):
+        log.debug(f"Requesting url: {url} with params: {api_params}")
+        retries = 0
+        delay = self.initial_delay
+        headers = {"Accept": "application/json"}
+        while retries <= self.max_retries:
```

Comment:
is this retry logic mostly copied from `ercot_api.py`? if so, we probably should abstract it out.

im okay if we want to leave that as a TODO./ followup task

Reply:
yeah copied over from ERCOT, marked in there with a TODO to consolidate and [made a ticket](https://www.notion.so/gridstatus/Consolidate-API-Handling-Methods-11ce835f42aa8043bfa0def723b4a45d?pvs=4).